gluetool/glue.py

```diff
@@ -1978,57 +2066,81 @@ def run_module(self, module_name, module_argv=None, actual_module_name=None, reg
+
+        return groups
+
+    def modules_descriptions(self, modules=None, groups=None):
+        # type: (Optional[ModuleRegistryType], Optional[List[str]]) -> str
+        """
+        Returns a string with modules and their descriptions.
+
+        :param dict(str, DiscoveredModule) modules: mapping with modules. If not set, current known modules
+            are used.
+        :param list(str) groups: if set, limit descriptions to modules belinging to the given groups.
+        :rtype: str
+        """
 
-        return sorted(self.modules)
+        modules = modules or self.modules
 
-    def module_list_usage(self, groups):
-        # type: (List[str]) -> str
+        as_groups = self.modules_as_groups(modules=modules)
 
-        """ Returns a string with modules description """
+        descriptions = []  # type: List[str]
 
         if groups:
-            usage = [
-                'Available modules in {} group(s)'.format(', '.join(groups))
+            descriptions += [
+                'Available modules in group(s) {}'.format(', '.join(groups))
             ]
+
         else:
-            usage = [
+            descriptions += [
                 'Available modules'
             ]
 
-        # get module list
-        plist = self.module_group_list()
-        if not plist:
-            usage.append('')
-            usage.append('  -- no modules found --')
+        def _add_no_modules():
+            # type: () -> None
+
+            descriptions.extend([
+                '',
+                '  -- no modules found --'
+            ])
+
+        if not as_groups:
+            _add_no_modules()
+
         else:
-            for group in sorted(plist):
+            descriptions += [
+                ''
+            ]
+
+            for group_name, group in sorted(as_groups.iteritems()):
                 # skip groups that are not in the list
-                # note that groups is [] if all groups should be shown
-                if groups and group not in groups:
+                # note that groups is None if all groups should be shown
+                if groups and group_name not in groups:
                     continue
-                usage.append('')
-                usage.append('%-2s%s' % (' ', group))
-                for key, val in sorted(plist[group].iteritems()):
-                    usage.append('%-4s%-32s %s' % ('', key, val))
 
-        return '\n'.join(usage)
+                group = as_groups[group_name]
 
-    def module_group_list(self):
-        # type: () -> Dict[str, Dict[str, str]]
+                if not group:
+                    _add_no_modules()
+                    continue
 
-        """ Returns a dictionary of groups of modules with description """
-        module_groups = {}  # type: Dict[str, Dict[str, str]]
+                for module_name, module in sorted(group.iteritems()):
+                    descriptions.append('%-4s%-32s %s' % ('', module_name, module.klass.description))
```

Comment:
What? This line needs definitely needs comment

Reply:
Agreed. Added a comment, and also switched to `format`.